typing/typecore.ml

```diff
@@ -5100,6 +5100,7 @@ and type_let ?check ?check_strict
 
   let (pat_list, exp_list, new_env, unpacks, _pvs) =
     with_local_level begin fun () ->
+      if existential_context = At_toplevel then Typetexp.TyVarEnv.reset ();
```

Comment:
@Octachron was displeased by the fact that `existential_restriction` is being for a different purpose. He volunteers to have two different types, `let_context` and `existential_restriction`, with a transformation from one to the other when we move from `type_let` to `type_pat`.